src/modules/rlm_mschap/rlm_mschap.c

```diff
@@ -560,82 +583,65 @@ static ssize_t mschap_xlat(UNUSED TALLOC_CTX *ctx, char **out, size_t outlen,
 		}
 
-		return strlen(*out);
+		fr_dcursor_append(out, vb);
+		return XLAT_ACTION_DONE;
 
 	/*
 	 * Return the NT-Hash of the passed string
 	 */
-	} else if (strncasecmp(fmt, "NT-Hash ", 8) == 0) {
-		char const *p;
-
-		p = fmt + 8;	/* 7 is the length of 'NT-Hash' */
-		if ((*p == '\0') || (outlen <= 32))
-			return 0;
-
-		fr_skip_whitespace(p);
+	} else if (strncasecmp(arg->vb_strvalue, "NT-Hash", 7) == 0) {
+		arg = fr_dlist_next(in, arg);
+		if ((!arg) || (arg->length == 0))
+			return XLAT_ACTION_FAIL;
 
-		if (mschap_nt_password_hash(buffer, p) < 0) {
+		if (mschap_nt_password_hash(buffer, arg->vb_strvalue) < 0) {
 			REDEBUG("Failed generating NT-Password");
 			*buffer = '\0';
-			return -1;
+			return XLAT_ACTION_FAIL;
 		}
 
-		fr_bin2hex(&FR_SBUFF_OUT(*out, (NT_DIGEST_LENGTH * 2) + 1), &FR_DBUFF_TMP(buffer, NT_DIGEST_LENGTH), SIZE_MAX);
-		(*out)[32] = '\0';
-		RDEBUG2("NT-Hash of \"known-good\" password: %s", *out);
-		return 32;
+		vb = fr_value_box_alloc_null(ctx);
+		fr_value_box_memdup_dbuff(ctx, vb, NULL, &FR_DBUFF_TMP(buffer, NT_DIGEST_LENGTH), NT_DIGEST_LENGTH, false);
+		RDEBUG2("NT-Hash of \"known-good\" password: %pV", vb);
+		fr_dcursor_append(out, vb);
+		return XLAT_ACTION_DONE;
 
 	/*
 	 * Return the LM-Hash of the passed string
 	 */
-	} else if (strncasecmp(fmt, "LM-Hash ", 8) == 0) {
-		char const *p;
-
-		p = fmt + 8;	/* 7 is the length of 'LM-Hash' */
-		if ((*p == '\0') || (outlen <= 32))
-			return 0;
-
-		fr_skip_whitespace(p);
-
-		smbdes_lmpwdhash(p, buffer);
-		fr_bin2hex(&FR_SBUFF_OUT(*out, (LM_DIGEST_LENGTH * 2) + 1), &FR_DBUFF_TMP(buffer, LM_DIGEST_LENGTH), SIZE_MAX);
-		(*out)[32] = '\0';
-		RDEBUG2("LM-Hash of %s = %s", p, *out);
-		return 32;
+	} else if (strncasecmp(arg->vb_strvalue, "LM-Hash", 7) == 0) {
+		arg = fr_dlist_next(in, arg);
+		if ((!arg) || (arg->length == 0))
+			return XLAT_ACTION_FAIL;
+
+		smbdes_lmpwdhash(arg->vb_strvalue, buffer);
+
+		vb = fr_value_box_alloc_null(ctx);
+		fr_value_box_memdup_dbuff(ctx, vb, NULL, &FR_DBUFF_TMP(buffer, LM_DIGEST_LENGTH), LM_DIGEST_LENGTH, false);
+		RDEBUG2("LM-Hash of %s = %pV", arg->vb_strvalue, vb);
+		fr_dcursor_append(out, vb);
+		return XLAT_ACTION_DONE;
 	} else {
-		REDEBUG("Unknown expansion string '%s'", fmt);
-		return -1;
+		REDEBUG("Unknown expansion string '%pV'", arg);
+		return XLAT_ACTION_FAIL;
 	}
 
-	if (outlen == 0) return 0; /* nowhere to go, don't do anything */
-
 	/*
 	 *	Didn't set anything: this is bad.
 	 */
 	if (!data) {
 		RWDEBUG2("Failed to do anything intelligent");
-		return 0;
-	}
-
-	/*
-	 *	Check the output length.
-	 */
-	if (outlen < ((data_len * 2) + 1)) {
-		data_len = (outlen - 1) / 2;
+		return XLAT_ACTION_FAIL;
 	}
 
-	/*
-	 *
-	 */
-	for (i = 0; i < data_len; i++) {
-		sprintf((*out) + (2 * i), "%02x", data[i]);
-	}
-	(*out)[data_len * 2] = '\0';
+	vb = fr_value_box_alloc_null(ctx);
+	fr_value_box_memdup_dbuff(ctx, vb, NULL, &FR_DBUFF_TMP(data, data_len), data_len, tainted);
```

Comment:
and here.  There's no need to use this function unless you're actually mduping from a dbuff